### DISCUSSION TITLE
WebAssembly support

### DISCUSSION THREAD

**Discussion by nsourligas** [cutelyst/cutelyst#338] [Q&A] [answered]
Hi,

I am currently working on a desktop QML application (Qt 5.15.0) and we were thinking to port it in web too using WebAssembly technology. But since we have some limitations due to sandbox nature of WebAssembly, an alternative solution for us it could be a REST API that cutelyst provides. So, do you know if it's possible to use cutelyst in a WASM application? And what about some security limitations of WebAssembly reported like this one [QTBUG-76350](https://bugreports.qt.io/browse/QTBUG-76350) ? Do they affect cutelyst compatibility with Qt WebAssembly?

Thanks!

**Comment by dantti** [comment, marked as answer]
There is no point in using cutelyst IN a WASM application, why would you want to listen as a server on the client's browser?
The limitation you pointed is about using RAW sockets with SSL, but calling an external REST API with HTTPS works just fine.
Cutelyst in WA should be the server part where you issue you HTTPS request and it works like if was PHP/JAVA/nodejs, with the advantage that it's Qt code that usually the one writing Qt desktop/WA/mobile app understand.
A side note is that it seems WA doesn't work with self signed certs, but I don't know the details just that a friend had this issue, which might not be an issue for you, I myself use WA with a cutelyst server using a valid cert and works just fine.

**Reply by nsourligas** [reply]
@dantti thank you for your reply! First, I would like to mention that maybe I posted my question a little bit recklessly without fully understanding the functionality of cutelyst framework since I haven't used it before. Consequently, for some reason I had a wrong theory in my mind that the cutelyst framework also covers the client part of a REST API which obviously doesn't happen.

So, of course, as you said there is no point in using cutelyst in a WASM application but only as a server (let's say a separate service) to issue https request and then handle separately by developing the client part in a WASM application. And this is the path we are going to follow in our project. 

Thanks for your answer and sorry if Ι wasn't clear enough due to the lack of cutelyst framework understanding.

**Reply by dantti** [reply]
Another possible use-case, is for the desktop version link to cutelyst and do the same networks requests it would if it was compiled as wasm, just using localhost on the same process this way you can share even more code, but it all depends on the use-case.

**Reply by nsourligas** [reply]
@dantti thanks for pointing out this possible use case, hmm nice I will keep in mind if it fits into any project depending on the case each time.

Btw really great job on cutelyst, asql and simple-mail projects, keep up the good work!
